### PR TITLE
Respect MAX_FILE_SIZE POST field in MultipartParser

### LINKED ISSUE #258: RequestBodyParserMiddleware should respect MAX_FILE_SIZE post value
The `MAX_FILE_SIZE` post value (cast insensitive) can be used to limit the maximum size for each following file upload. This is PHP's default behavior and it probably makes sense to mimic this here.

See also https://stackoverflow.com/questions/1381364/max-file-size-in-php-whats-the-point and https://github.com/php/php-src/blob/1c295d4a9ac78fcc2f77d6695987598bb7abcb83/main/rfc1867.c#L1053

This should not be confused with #257. In particular, this value can easily be spoofed and we have to ensure that this does not allow clients to bypass the server limits.

### PR DESCRIPTION
**Pull request by WyriHaximus**
Implements / closes #258

### CODE REVIEW

**Comment by WyriHaximus** [comment]
@clue @jsor This PR is read for review.

File: src/Io/MultipartParser.php [hunk 1]
```diff
@@ -134,14 +139,42 @@ private function parseFile($headers, $body)
...
+        $parsedBody = $this->request->getParsedBody();
+        if ($this->maxFileSize === null && is_array($parsedBody)) {
+            $postKeys = array_keys($parsedBody);
+            foreach ($postKeys as $key) {
+                if (strtoupper($key) === 'MAX_FILE_SIZE') {
+                    $this->maxFileSize = $parsedBody[$key];
+                    break;
+                }
+            }
+        }
```
Comment (jsor) [inline src/Io/MultipartParser.php]: If i remember correctly, it's not that easy. According to http://php.net/manual/en/features.file-upload.post-method.php#example-383, the field must precede the file input field.

The reason is, that you can have multiple `MAX_FILE_SIZE` hidden fields. Each field controls the max size of the following file fields.

```html
<input name="file1" type="file"> 
<input name="file2" type="file">
<!-- "file1" and "file2" will have no MAX_FILE_SIZE (no limit) -->

<input name="MAX_FILE_SIZE" value="1024" type="hidden">
<input name="file3" type="file">
<input name="file4" type="file">
<!-- "file3" and "file4" will have MAX_FILE_SIZE = 1024 bytes -->

<input name="MAX_FILE_SIZE" value="0" type="hidden">
<input name="file5" type="file">
<input name="file6" type="file">
<!-- "file5" and "file6" will have MAX_FILE_SIZE = 0 (no limit) -->
```

Comment (WyriHaximus) [reply src/Io/MultipartParser.php]: Updated the middleware to support this behavior :+

File: src/Io/MultipartParser.php [hunk 2]
```diff
@@ -154,6 +175,15 @@ private function parsePost($headers, $body)
...
                     $body
                 ));
+
+                if (strtoupper($matches[1]) === 'MAX_FILE_SIZE') {
+                    $this->maxFileSize = $body;
+
+                    if ($body == '0') {
+                        $this->maxFileSize = null;
+                    }
+                    break;
```
Comment (jsor) [inline src/Io/MultipartParser.php]: Why the `break` here?

Comment (WyriHaximus) [reply src/Io/MultipartParser.php]: Copy 🍝

Comment (WyriHaximus) [reply src/Io/MultipartParser.php]: Changed the :spaghetti: to 🍕

Comment (jsor) [reply src/Io/MultipartParser.php]: Yummy 🍽

File: src/Io/MultipartParser.php [hunk 3]
```diff
@@ -134,14 +139,30 @@ private function parseFile($headers, $body)
...
+        ));
+    }
+
+    private function parseUploadedFile($headers, $body)
+    {
+        $bodyLength = strlen($body);
+        if ($this->maxFileSize !== null && $bodyLength > $this->maxFileSize) {
+            return new UploadedFile(
+                Psr7\stream_for(''),
+                0,
```
Comment (clue) [inline src/Io/MultipartParser.php]: Couldn't find anything in PHP's documentation about this (http://php.net/manual/en/reserved.variables.files.php), but it's my understanding that we should still set the file size as given by the client here so consumers can show a more descriptive error message ("tried to upload X bytes, only Y allowed")?

File: src/Io/MultipartParser.php [hunk 4]
```diff
@@ -154,6 +175,14 @@ private function parsePost($headers, $body)
                     $matches[1],
                     $body
                 ));
+
+                if (strtoupper($matches[1]) === 'MAX_FILE_SIZE') {
+                    $this->maxFileSize = $body;
+
+                    if ($body == '0') {
+                        $this->maxFileSize = null;
+                    }
```
Comment (clue) [inline src/Io/MultipartParser.php]: Should probably cast to `int` above and use strict comparison here? This seems to be in line with how PHP parses this: https://github.com/php/php-src/blob/1c295d4a9ac78fcc2f77d6695987598bb7abcb83/main/rfc1867.c#L914

File: src/Io/MultipartParser.php (hunk 3 above)
Comment (WyriHaximus) [reply src/Io/MultipartParser.php]: SGTM :+1:

File: src/Io/MultipartParser.php (hunk 4 above)
Comment (WyriHaximus) [reply src/Io/MultipartParser.php]: SGTM :+1:

**Comment by WyriHaximus** [comment]
@clue added very minimal docs that will be expanded with the PR for #257

**Pull request merged**